### PR TITLE
TASK: Speed up matching in FrontendNodeRoutePartHandler

### PR DESCRIPTION
**Pull request by kdambekalns**
A fix for a performance issue: if a given node contains thousands of
child nodes, this implementation would load all these nodes into memory
in order to compare the current path segment with
`$node->getProperty('uriPathSegment')`.

### CODE REVIEW

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php [hunk 1]
```diff
@@ -378,40 +398,101 @@ protected function resolveRoutePathForNode(NodeInterface $node)
...
+              WHERE LOWER(CAST(nodedata.properties AS CHAR)) LIKE '%s'
+              AND nodedata.parentpathhash = '%s'
+              AND workspace IN (%s)
+              ORDER BY FIELD(workspace, %s) ASC
+              LIMIT 1
+        ";
+
+        $connection = $this->entityManager->getConnection();
+        foreach (explode('/', $relativeRequestPath) as $pathSegment) {
+            $statement = $connection->query(sprintf(
```
[the remark this answers is not in the document]
Comment (kdambekalns) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: Haha. Well, yo are right, but… I tried long and hard, but it didn't work.

Passing an array of strings (as needed for the `IN` clause) did not work, however hard I tried to understand what Doctrine expects there. And without that, I found it of no big use to do a prepared statement, when building (parts of) the SQL manually anyways.

I'd very much welcome a solution, though.

**Comment by kdambekalns** [comment]
BWT: This is marked as WIP since it

- does not cover PostgreSQL so far
- does not cover dimensions properly (i.e. in an optimized way)

**Comment by aertmann** [comment]
Curious if this PR has the same performance penalties as #672 does, when a node only has a few child nodes /cc @dimaip

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (kdambekalns) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: Except it didn't work for me when I tried. I'll gladly try again, though…

**Comment by kdambekalns** [comment]
Ah, thanks for the pointer to that one, @aertmann!

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (albe) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: If it really doesn't we should create an issue on doctrine-DBAL that their documentation is off :) (or their implementation broken)

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php [hunk 2]
```diff
@@ -378,40 +399,112 @@ protected function resolveRoutePathForNode(NodeInterface $node)
...
+            $workspaceNames[] = $workspace->getName();
+        });
+
+        $queryTemplate = 'SELECT path, parentPath
+              FROM neos_contentrepository_domain_model_nodedata AS nodedata
+              WHERE LOWER(CAST(nodedata.properties AS CHAR)) LIKE ?
+              AND nodedata.parentpathhash = ?
+              AND workspace IN (?)
+              ORDER BY FIELD(workspace, ?) ASC
+              LIMIT 1';
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: Changing the lines 457-463 to:
```
        $queryTemplate = 'SELECT path, parentPath
              FROM neos_contentrepository_domain_model_nodedata AS nodedata
              WHERE LOWER(CAST(nodedata.properties AS CHAR)) LIKE ?';
        if (!$siteNode->getContext()->isInvisibleContentShown()) {
            $queryTemplate .= '
                AND nodedata.hidden = 0
                AND nodedata.hidden = 0
                AND (nodedata.hiddenbeforedatetime IS NULL OR nodedata.hiddenbeforedatetime > NOW())
                AND (nodedata.hiddenafterdatetime IS NULL OR nodedata.hiddenafterdatetime < NOW())';
        }
        $queryTemplate .= '
              AND nodedata.parentpathhash = ?
              AND workspace IN (?)
              ORDER BY FIELD(workspace, ?) ASC
              LIMIT 1';
```
could solve the "hidden nodes are not filtered" issue

Comment (albe) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: Shouldn't the constraint that is indexed (parent path hash?) come first in order to help the query optimizer? I recall something like that. Could be outdated wisdom though

Comment (bwaidelich) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: yes, might be, good point

Comment (bwaidelich) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: =>
```
        $queryTemplate = 'SELECT path, parentPath
              FROM neos_contentrepository_domain_model_nodedata AS nodedata
              WHERE LOWER(CAST(nodedata.properties AS CHAR)) LIKE ?
              AND nodedata.parentpathhash = ?';
        if (!$siteNode->getContext()->isInvisibleContentShown()) {
            $queryTemplate .= '
                AND nodedata.removed = 0
                AND nodedata.hidden = 0
                AND (nodedata.hiddenbeforedatetime IS NULL OR nodedata.hiddenbeforedatetime > NOW())
                AND (nodedata.hiddenafterdatetime IS NULL OR nodedata.hiddenafterdatetime < NOW())';
        }
        $queryTemplate .= '
              AND workspace IN (?)
              ORDER BY FIELD(workspace, ?) ASC
              LIMIT 1';
```

FYI: Testing this change currently in a large project

Comment (albe) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: Can you verify the query execution plan with the change against your project database, i.e. `EXPLAIN SELECT ..` with different positions of the `parentpathhash` condition?

**Pull request closed without merging**